### DISCUSSION TITLE
formState.errors object is sometimes empty although field is not valid

### DISCUSSION THREAD

**Discussion by seyaobey** [react-hook-form/react-hook-form#4079] [Q&A]
Hello. I have added `react-hook-form` into my `react-native` app. In my `TextInput` component, I'm using `useController` to access `field` and `meta` keys to wire up the `onChange` and `onBlur` events and to access the value key. So here is (roughly) my component (MyTextInput):

```
export const MyTextInput = ({name, error, ...rest}) => {
    const { field, meta } = useController({name, control, rules}); // name = "username"
    return (
        <TextInput
              onChangeText={field.onChange}
              onBlur={field.onBlur}
              value={field.value}
              ref={field.ref}
              {...rest}
        />
   )
}
```

The `error` prop is passed from parent component from `useForm` like so:

```
const { formState, ... } = useForm();
...
return (
   <MyInputText name="username" error={formState.errors?.username} {...}/>  
)
```

Inside my text input component, I read `meta.isValid` to display an error helper text if validation rules are broken. What I observe is that sometimes while `meta.invalid = true`, the `formState.errors` object is empty. Thus I'm unable to display the error message. Why is the `formState.errors` empty (no error message) while `meta.invalid = true`? Is this a know bug or am I doing something wrong?

I hope my post makes sense. Thanks for your help.

**Comment by bluebill1049** [comment]
Did you miss this important message?

https://react-hook-form.com/api#formState

> Important: formState is wrapped with Proxy to improve render performance, so make sure you invoke or read it before render in 
order to enable the state update. This reduced re-render feature only applies to the Web platform due to a lack of support for Proxy in React Native.

**Reply by seyaobey** [reply]
@bluebill1049 thanks for the answer. I did not read that paragraph. But I did now, after you mentioned it. I tried to follow the suggestion of subscribing to `formState` changes through `useEffect`:

```
const {..., formState} = useForm()
useEffect(() => {
   if (formState.errors.username) {
      console.log('this works: I can read the field error');
   }
}, [formState])
```

But now I want to assign `formState.errors.username` to my local state and then pass it to my children TextInput components.
This is what I'm doing (simplified for sake of e.g):

```
const [errors, setErrors] = useState();

const {..., formState} = useForm():

useEffect(() => {
   if (formState.errors.username) {
      setErrors({
         username: formState.errors.username // <-- this does not work: Warning: Maximum update depth exceeded
      });
   }
}, [formState])
```
I could not figure out why infinite update loop is triggered. Is there a safe way to store `formState keys` to be reused later on?

Sorry sir, if I'm missing something...
I could not fully understand the part in the documentation stating that since `formState` is wrapped with `Proxy`, we should first `invoke or read it before render in order to enable the state update`. I could not find, from google search how to "invoke a proxy"...  

Also this part `This reduced re-render feature only applies to the Web platform due to a lack of support for Proxy in React Native.` Does that mean I should not use `formState` in a react-native app?

Sorry If I'm missing something sir, I just want to simply retrieve the error related to a field.

**Comment by seyaobey** [comment]
@bluebill1049 the way I solved it (not sure if this is the right solution), is to use `useMemo`

`const errors = useMemo(() => formState.errors, [formState]);`

It seems to work, no more infinite update loop and my child components correctly receive the error message. 
Waiting on your thoughts...
